xwayland/selection/dnd.c

```diff
@@ -334,9 +332,6 @@ void xwm_seat_handle_start_drag(struct wlr_xwm *xwm, struct wlr_drag *drag) {
 		wl_signal_add(&drag->events.destroy, &xwm->seat_drag_destroy);
 		xwm->seat_drag_destroy.notify = seat_handle_drag_destroy;
 
-		if (xwm->seat_drag_source_destroy.link.prev != NULL) {
-			wl_list_remove(&xwm->seat_drag_source_destroy.link);
-		}
```

Comment:
How about an assertion? I think it's fragile to assume `xwm->seat_drag_source_destroy` is not listening to any signal. A change to `seat_client_start_drag` may break this again.

Reply:
Nope, because when the data source is free'ed, all listeners are guaranteed to have been removed (otherwise they are corrupt). In other words, we require all listeners to remove themselves in the `destroy` signal handler.

Reply:
Yes, it works for now. But it may not work in the future as someone may change `seat_client_start_drag` (e.g. somebody need the source alive even after drag).

Reply:
In this case there can be multiple `seat->drag_source` at a time, and listeners shouldn't assume there's only one. In this case, `seat->drag_source` would probably need to be changed to be a list.

Another solution would be to keep only a single `seat->drag_source`, but to add a new `set_drag_source` signal to the seat, emitted when `seat->drag_source` is set. I'm not sure if this makes a lot of sense though.

Reply:
Alternatively, add some comment around `wlr_data_source_destroy(seat->drag_source);` to make sure anyone touching it is aware of the impact here.

Reply:
Done